zcncore/transaction_query.go

```diff
@@ -0,0 +1,401 @@
+
+			Logger.Debug(string(respBody))
+
+			return nil
+		})
+
+	errs := r.Wait()
+
+	if len(errs) > 0 {
+		return res, errs[0]
+	}
+
+	return res, nil
+
+}
+
+// FromAll query transaction from all sharders whatever it is selected or offline in previous queires, and return consensus result
+func (tq *TransactionQuery) FromAll(ctx context.Context, query string, handle QueryResultHandle) error {
+	if tq == nil || tq.max == 0 {
+		return ErrNoAvailableSharders
+	}
+
+	urls := make([]string, 0, tq.max)
+	for _, host := range tq.sharders {
+		urls = append(urls, tq.buildUrl(host, query))
+	}
+
+	r := resty.New()
+	r.DoGet(ctx, urls...).
+		Then(func(req *http.Request, resp *http.Response, respBody []byte, cf context.CancelFunc, err error) error {
+			res := QueryResult{
+				Content:    respBody,
+				Error:      err,
+				StatusCode: http.StatusBadRequest,
+			}
+			Logger.Debug(req.URL.String() + " " + resp.Status)
+
+			if resp != nil {
+				res.StatusCode = resp.StatusCode
+				Logger.Debug(string(respBody))
+			}
+
+			if handle != nil {
+				if handle(res) {
+
+					cf()
+				}
+			}
+
+			return nil
+		})
+
+	r.Wait()
+
+	return nil
+}
+
+// GetFastConfirmation get txn confirmation from a random online sharder
+func (tq *TransactionQuery) GetFastConfirmation(ctx context.Context, txnHash string) (*blockHeader, map[string]json.RawMessage, *blockHeader, error) {
+	var confirmationBlockHeader *blockHeader
+	var confirmationBlock map[string]json.RawMessage
+	var lfbBlockHeader blockHeader
+
+	// {host}/v1/transaction/get/confirmation?hash={txnHash}&content=lfb
+	result, err := tq.FromAny(ctx, tq.buildUrl("", TXN_VERIFY_URL, txnHash, "&content=lfb"))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	if result.StatusCode == http.StatusOK {
+
+		err = json.Unmarshal(result.Content, &confirmationBlock)
+		if err != nil {
+			Logger.Error("txn confirmation parse error", err)
+			return nil, nil, nil, err
+		}
+
+		// parse `confirmation` section as block header
+		confirmationBlockHeader, err = getBlockHeaderFromTransactionConfirmation(txnHash, confirmationBlock)
+		if err == nil {
+			return confirmationBlockHeader, confirmationBlock, nil, nil
+		}
+
+		Logger.Error("txn confirmation parse header error", err)
+
+		// parse `latest_finalized_block` section
+		lfbRaw, ok := confirmationBlock["latest_finalized_block"]
+		if !ok {
+			return confirmationBlockHeader, confirmationBlock, nil, err
+		}
+
+		err = json.Unmarshal([]byte(lfbRaw), &lfbBlockHeader)
+		if err == nil {
+			return confirmationBlockHeader, confirmationBlock, &lfbBlockHeader, ErrTransactionNotConfirmed
+		}
+
+		Logger.Error("round info parse error.", err)
+		return nil, nil, nil, err
+
+	}
+
+	return nil, nil, nil, thrown.Throw(ErrTransactionNotFound, strconv.Itoa(result.StatusCode))
+}
+
+func (tq *TransactionQuery) GetConsensusConfirmation(ctx context.Context, numSharders int, txnHash string) (*blockHeader, map[string]json.RawMessage, *blockHeader, error) {
```

Comment:
This method won't stop until all the tq.sharders have been checked, we will need to fix the cancel function in resty.DoGet() and Wait() methods. Otherwise the Consensus here would just not make sense.

Reply:
fixed